### ISSUE TITLE
Struct method passed to C casuses segmentation fault on call

### ISSUE THREAD

**Issue by odiroot** [vlang/v#20552]
**V doctor:**
```
V full version: V 0.4.4 ac2dcc2.0910401
OS: linux, "EndeavourOS Linux"
Processor: 12 cpus, 64bit, little endian, AMD Ryzen 5 PRO 4650U with Radeon Graphics

getwd: /home/odi/Temp/bugs
vexe: /home/odi/Repos/vlang/v
vexe mtime: 2024-01-10 21:43:04

vroot: OK, value: /home/odi/Repos/vlang
VMODULES: OK, value: /home/odi/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.43.0
Git vroot status: 0.4.4-13-g09104011 (41 commit(s) behind V master)
.git/config present: true

CC version: cc (GCC) 13.2.1 20230801
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

```

**What did you do?**
```
v -cc gcc -shared bug_method.v 
gcc bug_method.c bug_method.so -o bug_method -Wl,-rpath=.
./bug_method.so
```

`bug_method.v`:

```v
#flag -I.
#include "bug_method.h"

pub struct Foo {
	bar string
}

fn (f Foo) my_fun() &char {
	return ('Hello ' + f.bar).str
}

const f = Foo{
	bar: 'world'
}

@[typedef]
struct C.hello_t {
    say_hello fn() &char
}

@[export: "wrapper"]
fn wrapper() &C.hello_t {
    return &C.hello_t {
        say_hello: f.my_fun
    }
}
```

`bug_method.h`:
```c
typedef struct {
    char* (*say_hello)(void);
} hello_t;

```

`bug_method.c`:
```c
#include <stdio.h>
#include "bug_method.h"

extern hello_t* wrapper(void);

int main() {
    hello_t* h = wrapper();
    printf("%s\n", h->say_hello());

    return 0;
}
```

**What did you expect to see?**
C program with main print "Hello world".

**What did you see instead?**
```
[1]    83996 segmentation fault (core dumped)  ./bug_method.so
```

Basically, I wanted to demonstrate (somewhat awkwardly) that calling V struct methods from C code causes segmentation faults.

To do that I passed the pointer to my struct's method within a C struct and called it from `main` in C.  There's probably an easier way to show the effect but it escapes me for now.
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

**Comment by odiroot** [comment]
> You have to pass the struct as the first parameter. `(f Foo) my_fun() &char` is `char *my_fun(f Foo)` in C.

I guess that makes sense from V's PoV but sometimes you have no influence on the function signature, as it's defined by the C layer. 

In a sense, I expected the struct method to be a partial function -- expose the signature expected by C but wrapping the V struct reference.

**Comment by odiroot** [comment]
Let me give you a better real world example.

Here I'm setting up a C struct (`&C.clap_plugin_t`) before returning it to C layer. One of the function pointers (`init`) is actually a method on `ManualPlugin` instance.

https://github.com/mo-foss/vclap/blob/b774e9f0ab00adfcde6930d37115b7aa984a79a8/src/setup.v#L53

Method `init` on `ManualPlugin` is defined here:
https://github.com/mo-foss/vclap/blob/b774e9f0ab00adfcde6930d37115b7aa984a79a8/src/plugin.v#L8

The receiving C struct is wrapped in V here:
https://github.com/mo-foss/vclap/blob/b774e9f0ab00adfcde6930d37115b7aa984a79a8/src/c_abi.v#L36

And originaly in C header here:
https://github.com/mo-foss/vclap/blob/b774e9f0ab00adfcde6930d37115b7aa984a79a8/include/clap/plugin.h#L50

**Comment by SurmanPP** [comment]
If MinimalPlugin is an empty struct converting the methods to functions should fix it. However it still is bug that method types equal function types ignoring the struct they operate on.

**Comment by SurmanPP** [comment]
It was just a quick fix. It seems V is creating a closure with the `const f` however it fails at setting it up right.

**Comment by pierrec** [comment]
I think this is fixed as of `V 0.4.4 4163faf`. There was a fix for another issue when building in shared mode, which may also have fixed this one.
